### ISSUE TITLE
List of dependencies that should be updated from time to time

### ISSUE THREAD

**Issue by StrikerRUS** [microsoft/LightGBM#3763]
Here is a list of different places in LightGBM's GitHub repo where we specify some dependencies or helpers. Quite often we should specify a particular version of such software. And these versions are tend to obsolete with time. If you see that there is a newer version comparing to what we have, please feel free to propose a PR with updates.

- SHAs of commits of third-party dependencies in the [`external_libs` folder](https://github.com/microsoft/LightGBM/tree/master/external_libs). If there are some checkpoints (markers of stable versions) of these libraries, e.g. Releases at GitHub, the most recent stable release should be used (not the unstable latest `master` commit). If there are no such indicators of stable commit, use the latest available one. *How to [update GitHub submodules](https://stackoverflow.com/questions/5828324/update-git-submodule-to-latest-commit-on-origin).*
- SHAs of commits OpenCL libraries integration (same notes about versions as above):
1.https://github.com/microsoft/LightGBM/blob/f997a0692ca0f26740d2bdef2695c3e881d4e918/CMakeIntegratedOpenCL.cmake#L5-L6
2.https://github.com/microsoft/LightGBM/blob/f997a0692ca0f26740d2bdef2695c3e881d4e918/CMakeIntegratedOpenCL.cmake#L8-L9
3.https://github.com/microsoft/LightGBM/blob/f997a0692ca0f26740d2bdef2695c3e881d4e918/CMakeIntegratedOpenCL.cmake#L11-L12
- Visual Studio version for OpenCL libraries integration https://github.com/microsoft/LightGBM/blob/f997a0692ca0f26740d2bdef2695c3e881d4e918/CMakeIntegratedOpenCL.cmake#L66-L68
- MinGW w64 version at Appveyor https://github.com/microsoft/LightGBM/blob/f997a0692ca0f26740d2bdef2695c3e881d4e918/.appveyor.yml#L26
- [Read the Docs config file](https://github.com/microsoft/LightGBM/blob/master/.readthedocs.yml) (version of config file; adding new and removing deprecated config values)
- [`PublishBuildArtifacts` Azure Pipelines task](https://github.com/microsoft/LightGBM/blob/master/.vsts-ci.yml) (multiple occurrences, do search the file for `PublishBuildArtifacts` keyword)
- macOS image version at Azure Pipelines (**Attention!** Should be the **oldest** available, not the newest one) https://github.com/microsoft/LightGBM/blob/f997a0692ca0f26740d2bdef2695c3e881d4e918/.vsts-ci.yml#L144
- `DownloadBuildArtifacts` Azure Pipelines task https://github.com/microsoft/LightGBM/blob/f997a0692ca0f26740d2bdef2695c3e881d4e918/.vsts-ci.yml#L220
- `NuGetCommand` Azure Pipelines task https://github.com/microsoft/LightGBM/blob/f997a0692ca0f26740d2bdef2695c3e881d4e918/.vsts-ci.yml#L228
- `PublishBuildArtifacts` Azure Pipelines task https://github.com/microsoft/LightGBM/blob/f997a0692ca0f26740d2bdef2695c3e881d4e918/.vsts-ci.yml#L233
- `GitHubRelease` Azure Pipelines task https://github.com/microsoft/LightGBM/blob/f997a0692ca0f26740d2bdef2695c3e881d4e918/.vsts-ci.yml#L238
- CMake minimum required version and minimum supported versions of different compilers in [`CMakeLists.txt`](https://github.com/microsoft/LightGBM/blob/master/CMakeLists.txt) in case of adding new breaking changes in the CPP code of LightGBM
- Regenerate files in [Visual Studio solution folder](https://github.com/microsoft/LightGBM/tree/master/windows) in case of bumping minimum required version of **MSVC++**
- `Checkout repository` GitHub Action (multiple occurrences, do search in the [`.github/workflows` folder](https://github.com/microsoft/LightGBM/tree/master/.github/workflows) for `actions/checkout` keyword in all files)
- ...
- *I will continue filling this list slowly*

**Comment by StrikerRUS** [comment]
@jameslamb Please help to clarify. Should we monitor new releases for R 3.x stuff or they have been frozen forever?

**Comment by jameslamb** [comment]
> @jameslamb Please help to clarify. Should we monitor new releases for R 3.x stuff or they have been frozen forever?

No R 3.x resources will change.

Rtools for R 3.x is frozen forever. https://cran.r-project.org/bin/windows/Rtools/history.html

![image](https://user-images.githubusercontent.com/7608904/104822902-9a354900-580b-11eb-83a8-48f8ac99775a.png)

We also do not need to look for new versions of R itself. R doesn't have a concept of "long-term support" releases like other projects. So you won't see a bugfix that results in, say, R 4.1.1 and R 3.6.4 at the same time.

You can see the history of release dates at https://cran.r-project.org/bin/windows/base/old/.

**Comment by StrikerRUS** [comment]
I'm so sorry for my English, but I can read this answer in two ways. Which is correct? 😄 

> No,     R 3.x resources **will** change.

> **No R 3.x resources** will change.

____

> Rtools for R 3.x is frozen forever.

Ah, I see, thanks! Removed from the list.

> You can see the history of release dates at https://cran.r-project.org/bin/windows/base/old/.

I see there are

```
R 4.0.3 (October, 2020)
R 4.0.2 (June, 2020)
R 4.0.1 (June, 2020)
R 4.0.0 (April, 2020)
```

What about `4.0.4`? Why we shouldn't wait for it?

**Comment by jameslamb** [comment]
Oh sorry, let me rephrase.

* there will never be a new R 3.x release
* there will never be a new Rtools 3.x release

For 4.x series, there will continue to be new releases. They tend to do 4 or 5 per year. The next release of R will be R 4.1.0 (). That might or might not involve a new Rtools 4.x release. I use this mirror of the R source (R development is not actually done on GitHub) to keep track: https://github.com/wch/r-source/blob/trunk/VERSION

**Comment by StrikerRUS** [comment]
Thank you very much!
Then I'm removing R 3.x from the list as well.

> The next release of R will be R 4.1.0. That might or might not involve a new Rtools 4.x release.

Got it, great! Then I think we can have them both in the list. 

> They tend to do 4 or 5 per year.

Quite frequent. For instance,
> Visual Studio version for OpenCL libraries integration

VS releases have 2-year cycle, but I still find it useful to have VS mentioned here.

**Comment by jameslamb** [comment]
I want to note here that R 4.1.0 is planned for release on May 18, 2021.

https://twitter.com/pdalgd/status/1383008275545919488

You can see a full list of changes in that version at https://stat.ethz.ch/R-manual/R-devel/doc/html/NEWS.html. There is only one that looks immediately relevant to `{lightgbm}`, although it's hard to tell since some of the changes are deep in the internals of R and `{lightgbm}` may indirectly rely on them.

> The default C++ standard has been changed to C++14 where available (which it is on all currently checked platforms): if not (as before) C++11 is used if available otherwise C++ is not supported.

> Packages which specify C++11 will still be installed using C++11.

> Either specify C++11 (see ‘Writing R Extensions’) or modernize the code and if needed specify C++14

I don't think `{lightgbm}` needs to worry about this, because it already declares its use of C++11 in the way that is recommended by "Writing R Extensions"

* guidelines saying to set `CXX_STD` in `Makevars`: https://cran.r-project.org/doc/manuals/r-release/R-exts.html#Using-C_002b_002b-code
    - `Makevars.in`: https://github.com/microsoft/LightGBM/blob/7ea2bc4dbca53c03eaae2aa2b22f9b30447caea4/R-package/src/Makevars.in#L1
    - `Makevars.win.in`: https://github.com/microsoft/LightGBM/blob/7ea2bc4dbca53c03eaae2aa2b22f9b30447caea4/R-package/src/Makevars.win.in#L1

I don't think that any action is required at this time to prepare for R 4.1.0. But that release will be a big one so I just wanted to note that it's coming so we're prepared.

**Comment by jameslamb** [comment]
This project's Windows R-package CI jobs will need to update to a new RTools toolchain, `RTools43`, once R 4.3 is released.

docs: https://cran.r-project.org/bin/windows/Rtools/rtools43/rtools.html

announcement on the `R-pkg-devel` mailing list: https://stat.ethz.ch/pipermail/r-package-devel/2023q1/008780.html

**Comment by StrikerRUS** [comment]
Looks like this issue haven't been updated from more than 1.5 year. I guess a lot of things have been changed.

Should we close this issue?
@jameslamb

**Comment by jameslamb** [comment]
Many of the things mentioned in this issue have been updated in the last year and a half, but I hadn't started from this issue and so did not remember to link it.

The list is a very useful list of all the places that could be updated and I'd hoped we would get outside contributors helping with this via the `good first issue` label, but:

* you and I are the only people who've linked back to this issue or edited it in the 3.5 years since it was opened
* it takes some effort to keep the links and list of items up to date

Despite that.... **I vote that we keep it open**. I will try to do a better job of linking back to this in version updates, and can spend some time updating the list and links in the issue description.

I also think there are some opportunities to use more automation for managing versions and to centralize version numbers in fewer places.

For example:

* @borchero set up `dependabot` for GitHub Actions versions in #6449
* I'm very interested in exploring Renovate, which can generically look at things like PyPI releases and git tags and suggest changes, and where you can centralize those versions in a file
  - https://docs.renovatebot.com/
  - example from RAPIDS:
    - renovate updating a file: https://github.com/rapidsai/ci-imgs/blob/1426d6b13d862c1427711c97d8aa46d702501253/versions.yaml#L10
    - that file's content being passed through as a docker build arg (just as an example): https://github.com/rapidsai/ci-imgs/blob/1426d6b13d862c1427711c97d8aa46d702501253/ci-conda.Dockerfile#L9

**Comment by StrikerRUS** [comment]
@jameslamb Thanks a lot for your position! Renovate looks interesting! I remember I saw it somewhere in the past but didn't give it a try, unfortunately...

**Comment by borchero** [comment]
@jameslamb without reading through the entirety of this issue:

> SHAs of commits of third-party dependencies

we could also automate this using dependabot just like GH actions :smile:

**Comment by jameslamb** [comment]
> *we could also automate this using dependabot just like GH actions*

Oh that would be excellent!

I'd support a PR that does that. At this point I'm very confident in this project's CI to catch issues with new versions of those dependencies, and it would be great to integrate bugfixes and performance improvements more often.

**Comment by jameslamb** [comment]
We had one extra spot in pinned issues, so I decided to pin this for now so it'll be more visible from https://github.com/microsoft/LightGBM/issues

**Comment by Shobanasrj** [comment]
Hi, I’d love to work on this issue! Could you provide any additional guidance on how you’d like the dependency list to be maintained?

**Comment by jameslamb** [comment]
Thanks! Just look into any of those links that reference things you're familiar with, look for the corresponding things on tbe latest state of `master`, and propose a pull request updating one of them.
